File: nwerc2023/joggingtour/submissions/accepted/jeroen.py
#!/usr/bin/env python3
import math

INF = 1e99

n = int(input())
pts = []
for i in range(n):
    [x,y] = input().split(' ')
    pts.append((int(x),int(y)))


def dist(i,j,angle):
    (vx,vy) = (pts[i][0] - pts[j][0], pts[i][1] - pts[j][1])
    rx = vx * math.cos(angle) + vy * math.sin(angle)
    ry = -vx * math.sin(angle) + vy * math.cos(angle)
    return abs(rx) + abs(ry)

def givenAngle(angle):
    # Dp state: places to visit, current place
    pw = 1 << n
    dp = [ [0] * n for _ in range(pw) ]
    for i in range(1,pw):
        for j in range(n):
            dp[i][j] = INF;
            for k in range(n):
                if (i & (1<<k)) == 0: continue
                if j == k: continue
                dp[i][j] = min(dp[i][j], dp[i^(1<<k)][k] + dist(j,k,angle))

    # Find the right starting place
    best = INF
    for i in range(n):
        best = min(best, dp[(1<<n)-(1<<i)-1][i])
    return best

best = None

for i in range(n):
    for j in range(i):
        angle = math.atan2(pts[i][0] - pts[j][0], pts[i][1] - pts[j][1])
        cur = givenAngle(-angle)
        if not best or cur < best:
            best = cur

print(best)

File: nwerc2023/joggingtour/submissions/accepted/jeroen.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define MIN(a,b) ((a) < (b) ? (a) : (b))
#define ABS(a) ((a) < 0 ? -(a) : (a))

#define MAXN 12

int x[MAXN];
int y[MAXN];
int n;

double dp[1<<MAXN][MAXN];
double dist[MAXN][MAXN];

double distance(double angle, int i, int j) {
  double dx = x[j] - x[i];
  double dy = y[j] - y[i];
  double x = dx * cos(angle) - dy * sin(angle);
  double y = dx * sin(angle) + dy * cos(angle);
  return ABS(x) + ABS(y);
}

double withAngle(double angle) {
  for(int i = 0; i < n; i++)
    for(int j = i + 1; j < n; j++)
      dist[i][j] = dist[j][i] = distance(-angle, i, j);
  
  dp[0][0] = 0;
  for(int i = 1; i < (1<<n); i++) {
    for(int j = 0; j < n; j++) {
      dp[i][j] = 1e99;
      for(int k = 0; k < n; k++) {
	if((i & (1<<k)) == 0) continue;
	if(j == k) continue;
	dp[i][j] = MIN(dp[i][j], dp[i^(1<<k)][k] + dist[j][k]);
      }
    }
  }
  double ret = 1e99;
  for(int i = 0; i < n; i++)
    ret = MIN(ret, dp[(1<<n)-(1<<i)-1][i]);
  return ret;
}

int main() {
  scanf("%d\n", &n);
  for(int i = 0; i < n; i++)
    scanf("%d %d\n", &x[i], &y[i]);

  double best = 1e99;

  for(int i = 0; i < n; i++) {
    for(int j = i + 1; j < n; j++) {
      double angle = atan2(y[j] - y[i], x[j] - x[i]);
      double cur = withAngle(angle);
      if(cur < best) best = cur;
    }
  }

  printf("%lf\n", best);

  return 0;
}
